src/ripple/app/tx/impl/URIToken.cpp

```diff
@@ -0,0 +1,819 @@
+    }
+
+    uint32_t flags = ctx_.tx.getFlags();
+
+    std::shared_ptr<SLE> sleU;
+    std::optional<AccountID> issuer;
+    std::optional<AccountID> owner;
+    std::optional<STAmount> saleAmount;
+    std::optional<AccountID> dest;
+    std::optional<Keylet> kl;
+    std::shared_ptr<SLE> sleOwner;
+
+    if (tt != ttURITOKEN_MINT)
+    {
+        kl = Keylet{ltURI_TOKEN, ctx_.tx.getFieldH256(sfURITokenID)};
+        sleU = view().peek(*kl);
+
+        if (!sleU)
+            return tecNO_ENTRY;
+
+        if (sleU->getFieldU16(sfLedgerEntryType) != ltURI_TOKEN)
+            return tecNO_ENTRY;
+
+        owner = (*sleU)[sfOwner];
+        issuer = (*sleU)[sfIssuer];
+        saleAmount = (*sleU)[~sfAmount];
+        dest = (*sleU)[~sfDestination];
+
+        if (*owner == account_)
+            sleOwner = sle;
+        else
+            sleOwner = view().peek(keylet::account(*owner));
+
+        if (!sleOwner)
+        {
+            JLOG(j.warn()) << "Malformed transaction: owner of URIToken is not "
+                              "in the ledger.";
+            return tecNO_ENTRY;
+        }
+    }
+
+    switch (tt)
+    {
+        case ttURITOKEN_MINT: {
+            kl = keylet::uritoken(account_, ctx_.tx.getFieldVL(sfURI));
+            if (view().exists(*kl))
+                return tecDUPLICATE;
+
+            sleU = std::make_shared<SLE>(*kl);
+            sleU->setAccountID(sfOwner, account_);
+            sleU->setAccountID(sfIssuer, account_);
+            sleU->setFieldVL(sfURI, ctx_.tx.getFieldVL(sfURI));
+
+            if (ctx_.tx.isFieldPresent(sfDigest))
+                sleU->setFieldH256(sfDigest, ctx_.tx.getFieldH256(sfDigest));
+
+            if (flags & tfBurnable)
+                sleU->setFlag(tfBurnable);
+
+            auto const page = view().dirInsert(
+                keylet::ownerDir(account_), *kl, describeOwnerDir(account_));
+
+            JLOG(j_.trace())
+                << "Adding URIToken to owner directory " << to_string(kl->key)
+                << ": " << (page ? "success" : "failure");
+
+            if (!page)
+                return tecDIR_FULL;
+
+            sleU->setFieldU64(sfOwnerNode, *page);
+            view().insert(sleU);
+
+            adjustOwnerCount(view(), sle, 1, j);
+            return tesSUCCESS;
+        }
+
+        case ttURITOKEN_CANCEL_SELL_OFFER: {
+            sleU->makeFieldAbsent(sfAmount);
+            if (sleU->isFieldPresent(sfDestination))
+                sleU->makeFieldAbsent(sfDestination);
+            view().update(sleU);
+            return tesSUCCESS;
+        }
+
+        case ttURITOKEN_BUY: {
+            if (account_ == *owner)
+            {
+                // this is a clear operation
+                sleU->makeFieldAbsent(sfAmount);
+                if (sleU->isFieldPresent(sfDestination))
+                    sleU->makeFieldAbsent(sfDestination);
+                view().update(sleU);
+                return tesSUCCESS;
+            }
```

Comment:
it's a bit surprising to me that buying my own token has the same behavior as canceling my sell offer. what's the rationale?

Reply:
What behavior would you expect or anything specific that made you question it?

From an accounting/reconciliation POV I can see your point as this might need to be recorded for those purposes.

Reply:
It's a valid purchase it just results in 0 movement of objects or funds.

Reply:
i think a comment to clarify would be helpful :)

Reply:
this also allows me to buy it from myself even though i haven't listed it for sale. it's really confusing to me and i'm not sure that that means it is a valid purchase. i could see it in the case where i have listed it for sale and then i attempt to buy it from myself